File: 0475-heaters/0475-heaters.cpp
class Solution {
public:
    int nearestheaterdistance(vector<int> &heaters,int house){
        int start = 0;
        int end = heaters.size()-1;
        int ans = INT_MAX;
        while(start<=end){
            int mid = start+(end-start)/2;
            if(heaters[mid]==house) return 0;

            ans = min(ans,abs(house-heaters[mid]));
            if(heaters[mid] > house){
                end = mid-1;
            }else{
                start = mid+1;
            }
        }
        return ans;
    }
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        sort(heaters.begin(),heaters.end());
        int distance = 0;
        for(auto house:houses){
            int dis = nearestheaterdistance(heaters,house);
            distance = max(distance,dis);
        }
        return distance;
    }
};

File: 0475-heaters/0475-heaters.java
class Solution {
    private int min(int a,int b){
        if(a <= b ) return a;
        return b;
    }
    private int closestheaterdistance(int[] heaters,int house){
        int start = 0;
        int end = heaters.length-1;
        int ans = Integer.MAX_VALUE;
        while(start<=end){
            int mid = start + (end-start)/2;
            ans = min(ans,Math.abs(heaters[mid]-house));

            if(heaters[mid]>house){
                end = mid-1;
            }else{
                start = mid+1;
            }
        
        }
        return ans;
    }
    public int findRadius(int[] houses, int[] heaters) {
        int distance = 0;
        Arrays.sort(heaters);
        for(int it:houses){
            distance = Math.max(distance,closestheaterdistance(heaters,it));
        }
        return distance;
    }
}